lib/openhab/dsl/items/semantics.rb

```diff
@@ -0,0 +1,179 @@
+# frozen_string_literal: true
+
+module OpenHAB
+  module DSL
+    # Module for implementing semantics helper methods on [GenericItem]
+    #
+    # Wraps https://www.openhab.org/javadoc/latest/org/openhab/core/model/script/actions/semantics,
+    # as well as adding a few additional convenience methods.
+    # Also includes Classes for each semantic tag.
+    # See https://github.com/openhab/openhab-core/blob/main/bundles/org.openhab.core.semantics/model/SemanticTags.csv
+    module Semantics
+      SemanticsAction = org.openhab.core.model.script.actions.Semantics
+      private_constant :SemanticsAction
+
+      # import all the semantics constants
+      [org.openhab.core.semantics.model.point.Points,
+       org.openhab.core.semantics.model.property.Properties,
+       org.openhab.core.semantics.model.equipment.Equipments,
+       org.openhab.core.semantics.model.location.Locations].each do |parent_tag|
+        parent_tag.stream.for_each do |tag|
+          const_set(tag.simple_name.to_sym, tag.ruby_class)
+        end
+      end
+
+      # put ourself into the global namespace, replacing the action
+      ::Semantics = self # rubocop:disable Naming/ConstantName
+
+      # Checks if this Item is a Location
+      # @return [true, false]
+      def location?
+        SemanticsAction.location?(self)
+      end
+
+      # Checks if this Item is an Equipment
+      # @return [true, false]
+      def equipment?
+        SemanticsAction.equipment?(self)
+      end
+
+      # Checks if this Item is a Point
+      # @return [true, false]
+      def point?
+        SemanticsAction.point?(self)
+      end
+
+      # Checks if this Item has any semantic tags
+      # @return [true, false]
+      def semantic?
+        !!semantic_type
+      end
+
+      # Gets related Equipment [GenericItem] an [Item] belongs to.
+      #
+      # If this Item is a Location and a [GroupItem], it will return
+      # an array of all Equipment items within the Location. Otherwise
+      # it returns the single related Equipment Item this Item belongs
+      # to (or none, if it doesn't belong to any).
+      #
+      # Note that an Equipment Item will return itself.
+      #
+      # @return [GenericItem, Array<GenericItem>, nil]
+      def equipment
+        return SemanticsAction.get_equipment(self) unless location? && is_a?(GroupItem)
+
+        members.select(&:equipment?)
+      end
+
+      # Returns the sub-class of [Equipment] this Item is tagged with.
+      # @return [Class]
+      def equipment_type
+        SemanticsAction.get_equipment_type(self)&.ruby_class
+      end
+
+      # Gets the related Location Item of this Item.
+      # @return [GenericItem]
+      def location
+        SemanticsAction.get_location(self)
+      end
+
+      # Returns the sub-class of [Location] this Item is tagged with.
+      # @return [Class]
+      def location_type
+        SemanticsAction.get_location_type(Self)&.ruby_class
```

Comment:
Is `Self` not a typo?

Reply:
Yes, indeed it is. This is why I need tests :)

Reply:
I'm curious, mainly because I am still not familiar with `ruby_class`. Some of the calls in other methods didn't convert the returned object into ruby_class like this one, and I'm wondering what the difference.